File: packages/provide-uterm-server/src/provide/uterm/server/connectors/registry.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""Connector self-registration registry."""

from __future__ import annotations

import contextlib
import importlib
from typing import TYPE_CHECKING, Any, Protocol

if TYPE_CHECKING:
    from provide.uterm.server.connectors.base import SessionConnector


class SessionConnectorFactory(Protocol):
    def __call__(self, session_id: str, display_name: str, config: dict[str, Any], /) -> SessionConnector: ...


_registry: dict[str, SessionConnectorFactory] = {}
_BUILTIN_CLASSES: dict[str, tuple[str, str]] = {
    "shell": ("provide.uterm.server.connectors.shell", "ShellSessionConnector"),
    "ssh": ("provide.uterm.server.connectors.ssh", "SshSessionConnector"),
    "telnet": ("provide.uterm.server.connectors.telnet", "TelnetSessionConnector"),
    "websocket": ("provide.uterm.server.connectors.websocket", "WebSocketSessionConnector"),
}


def register_connector(name: str, cls: SessionConnectorFactory) -> None:
    """Register a connector class under a type name."""
    _registry[name] = cls


def build_connector(
    session_id: str,
    display_name: str,
    connector_type: str,
    config: dict[str, Any],
) -> SessionConnector:
    """Instantiate a connector by type name. Raises ValueError for unknown types."""
    cls = _registry.get(connector_type)
    if cls is None:
        builtin = _BUILTIN_CLASSES.get(connector_type)
        if builtin is not None:
            module_name, class_name = builtin
            with contextlib.suppress(ImportError):
                module = importlib.import_module(module_name)
                register_connector(connector_type, getattr(module, class_name))
            cls = _registry.get(connector_type)
    if cls is None:
        raise ValueError(f"unsupported connector_type: {connector_type!r}")
    return cls(session_id, display_name, config)


def registered_types() -> frozenset[str]:
    """Return the set of currently registered connector type names."""
    return frozenset(_registry)

File: packages/provide-uterm-ts/src/connectors/registry.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * Connector self-registration.
 *
 * Port of the Python module `provide.uterm.server.connectors.registry`.
 *
 * An unknown type is a refusal rather than a default: a session created with
 * a typo must not silently land on some other transport.
 */

import type { SessionConnector } from "./base.ts";
import { ShellSessionConnector } from "./shell.ts";

/** Builds a connector for one session. */
export type SessionConnectorFactory = (
  sessionId: string,
  displayName: string,
  config: Record<string, unknown>,
) => SessionConnector;

/**
 * The types the reference ships with.
 *
 * The network connectors are named here so the set matches the reference even
 * though only the reference connector is built in this port so far.
 */
export const BUILTIN_CONNECTOR_TYPES: readonly string[] = ["shell", "ssh", "telnet", "websocket"];

const registry = new Map<string, SessionConnectorFactory>([
  ["shell", (sessionId, displayName, config) => new ShellSessionConnector(sessionId, displayName, config)],
]);

/**
 * Register a factory under a type name.
 *
 * A later registration replaces an earlier one: that is how a deployment
 * substitutes its own transport for a built-in.
 */
export function registerConnector(name: string, factory: SessionConnectorFactory): void {
  registry.set(name, factory);
}

/**
 * Build a connector by type name.
 *
 * @throws {Error} On a type nothing has registered.
 */
export function buildConnector(
  sessionId: string,
  displayName: string,
  connectorType: string,
  config: Record<string, unknown>,
): SessionConnector {
  const factory = registry.get(connectorType);
  if (factory === undefined) {
    throw new Error(`unsupported connector_type: '${connectorType}'`);
  }
  return factory(sessionId, displayName, config);
}

/** The type names currently registered. */
export function registeredTypes(): ReadonlySet<string> {
  return new Set(registry.keys());
}
